rkt/image_extract.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2015 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/rkt/pkg/fileutil"
+	"github.com/coreos/rkt/store"
+
+	"github.com/coreos/rkt/Godeps/_workspace/src/github.com/spf13/cobra"
+)
+
+var (
+	cmdImageExtract = &cobra.Command{
+		Use:   "extract IMAGE OUTPUT_ACI_FILE",
+		Short: "extract a stored image to an ACI file",
+		Long:  `IMAGE should be a string referencing an image: either a hash or an image name.`,
+		Run:   runWrapper(runImageExtract),
+	}
+	flagRootfsOnly bool
+	flagOverwrite  bool
+)
+
+func init() {
+	cmdImage.AddCommand(cmdImageExtract)
+	cmdImageExtract.Flags().BoolVar(&flagRootfsOnly, "rootfs-only", false, "extract rootfs only")
+	cmdImageExtract.Flags().BoolVar(&flagOverwrite, "overwrite", false, "overwrite output directory")
+}
+
+func runImageExtract(cmd *cobra.Command, args []string) (exit int) {
+	if len(args) != 2 {
+		cmd.Usage()
+		return 1
+	}
+
+	s, err := store.NewStore(globalFlags.Dir)
+	if err != nil {
+		stderr("image extract: cannot open store: %v", err)
+		return 1
+	}
+
+	key, err := getKeyFromAppOrHash(s, args[0])
+	if err != nil {
+		stderr("image extract: %v", err)
+		return 1
+	}
+
+	if err := s.RenderTreeStore(key, false); err != nil {
+		stderr("image extract: error rendering ACI: %v", err)
+		return 1
+	}
+	if err := s.CheckTreeStore(key); err != nil {
+		stderr("image extract: warning: tree cache is in a bad state. Rebuilding...")
+		if err := s.RenderTreeStore(key, true); err != nil {
+			stderr("image extract: error rendering ACI: %v", err)
+			return 1
+		}
+	}
+
+	if _, err := os.Stat(args[1]); err == nil {
+		if !flagOverwrite {
+			stderr("image extract: output directory exists (try --overwrite)")
+			return 1
+		}
+
+		if err := os.RemoveAll(args[1]); err != nil {
+			stderr("image extract: error removing existing output dir: %v", err)
+			return 1
+		}
+	}
+
+	if err := os.MkdirAll(args[1], 0755); err != nil {
+		stderr("image extract: error creating out directory: %v", err)
+		return 1
+	}
+
+	rootfsOutDir := args[1]
+	if !flagRootfsOnly {
+		rootfsOutDir = filepath.Join(rootfsOutDir, "rootfs")
+
+		manifest, err := s.GetImageManifest(key)
+		if err != nil {
+			stderr("image extract: error getting manifest: %v", err)
+			return 1
+		}
+
+		mb, err := json.Marshal(manifest)
```

Comment:
Probably not an issue to be addressed in this PR, but it's a bit troubling that this is nondeterministic, I think after doing an extract and then an `actool build` the image hash is going to be different :/
(kinda related: #544 , appc/spec#199)